js/tests/unit/modal.js

```diff
@@ -658,4 +658,17 @@ $(function () {
       .bootstrapModal('show')
       .bootstrapModal('hide')
   })
+
+  QUnit.test('should dispose modal', function (assert) {
+    assert.expect(1)
+    var done = assert.async()
+
+    var $modal = $('<div id="modal-test"/>')
+      .bootstrapModal('show')
+      .bootstrapModal('hide')
```

Comment:
I don't understand the need to use `show/hide` methods here 🤔 

```js
var $modal = $('<div id="modal-test"/>').bootstrapModal()

assert.ok(typeof $modal.data('bs.modal') !== 'undefined', 'modal created')

$modal.bootstrapModal('dispose')

assert.ok(typeof $modal.data('bs.modal') === 'undefined', 'modal data object was disposed')
```

If you want to go further you can use a SinonJS spy to be sure `dispose` is called

Reply:
From start `bs.modal` is undefined, then is set on `show`, then is removed again on `hide`

Reply:
Disposing modal in `show` state is ambiguous. All listeners are removed, and you're stuck with an open modal on page.

Reply:
![image](https://user-images.githubusercontent.com/1881266/47266138-047eb680-d53b-11e8-8af4-bce7bbbb7ba5.png)

Reply:
The modal instance is created here: https://github.com/twbs/bootstrap/blob/94a0144b2cfa5d50b2976295da9d07465acb22ad/js/src/modal.js#L509
And me remove the data on `dispose` here: https://github.com/twbs/bootstrap/blob/94a0144b2cfa5d50b2976295da9d07465acb22ad/js/src/modal.js#L199

No need to show or hide

Reply:
See my CodePen: https://codepen.io/Johann-S/pen/jepKdY

Reply:
let's wait till GitHub will work again :smile:

Reply:
To `dispose` a modal you have to `show` it first.  
Otherwise `$('#myModal').data('bs.modal')` will be `undefined`, and `dispose` won't have any effect.  
If you `dispose` it in `show` state you're blocked with an open modal on page.  

![image](https://user-images.githubusercontent.com/1881266/47369328-18b5e580-d6ec-11e8-88a1-d77a3b286b8f.png)

Also if you dispose a modal in transition state (with fade `show -> shown`), you get a TypeError  

![image](https://user-images.githubusercontent.com/1881266/47369243-f1f7af00-d6eb-11e8-9aef-582e0db6c738.png)

https://codepen.io/iamandrewluca/pen/aRRbXb

This is why I `show/hide` modal, then `dispose`

Reply:
My CodePen show the opposite please take time to Open it with your devtools open

Reply:
Your CodePen works as it should be. There you only `show` the modal.  
Try to dispose it after `show`

Reply:
Yep but what I'm showing, it's, you don't have to call `show` to create an instance of our modal.
Because my CodePen JS is just:

```js
$(function () {
  $('#myModal').modal()
  
  console.warn($('#myModal').data('bs.modal'))
})

Reply:
You mean I have just to call `$('#myModal').modal()` ?
This 2 are the same
```js
$('#myModal').modal() // will call `show` method
$('#myModal').modal('show') // will call `show` method
```
On tests page I get an error if I don't hide modal before dispose, when trying to click on something

```js
QUnit.test('should dispose modal', function (assert) {
  assert.expect(1)
  var done = assert.async()

  var $modal = $('<div id="modal-test"/>')
    .bootstrapModal()
    .bootstrapModal('dispose')

  assert.ok(typeof $modal.data('bs.modal') === 'undefined', 'modal data object was disposed')
  done()
})
```
![image](https://user-images.githubusercontent.com/1881266/47417903-adb6ee00-d781-11e8-8ca9-67c4e47a00f1.png)

Also if I don't hide, body will have the padding added when there is overflow.

Reply:
Tested that : 
```js
  QUnit.test('should dispose modal', function (assert) {
    assert.expect(2)
    var $el = $('<div id="modal-test"/>')
    $el.bootstrapModal()

    assert.ok($el.data('bs.modal') !== undefined)

    $el.bootstrapModal('dispose')

    assert.ok($el.data('bs.modal') === undefined)
  })
```

and it works fine on our unit tests

Reply:
Tried to click any input on the tests page? you will get an error, and also body padding is added.

Reply:
That's not the purpose here, we test if our `dispose` methods do the job that's all, so we just have to check if the data is removed and if the listeners too.

I did the first check (check if the data is removed).
We did that on other tests for our Modal and it's fine.

Reply:
If you want we can discuss about that on the Bootstrap Slack 😉